File: goblin-cache/src/main/java/org/goblinframework/cache/util/CacheExpirationCalculator.java
package org.goblinframework.cache.util;

import org.goblinframework.cache.core.annotation.GoblinCacheExpiration;
import org.goblinframework.core.util.DateUtils;
import org.jetbrains.annotations.NotNull;

abstract public class CacheExpirationCalculator {

  public static int expirationInSeconds(@NotNull GoblinCacheExpiration annotation) {
    if (!annotation.enable()) {
      throw new IllegalArgumentException("@UtopiaCacheExpiration is not enabled");
    }
    return expirationInSeconds(annotation.policy(), annotation.value());
  }

  public static int expirationInSeconds(@NotNull GoblinCacheExpiration.Policy policy, int value) {
    switch (policy) {
      case FIXED: {
        return Math.max(0, value);
      }
      case TODAY: {
        return DateUtils.getCurrentToDayEndSecond();
      }
      case THIS_WEEK: {
        return DateUtils.getCurrentToWeekEndSecond();
      }
      case THIS_MONTH: {
        return DateUtils.getCurrentToMonthEndSecond();
      }
      default: {
        throw new UnsupportedOperationException();
      }
    }
  }
}

File: goblin-cache/src/main/java/org/goblinframework/cache/core/_CacheExpirationCalculator.kt
package org.goblinframework.cache.core

import org.goblinframework.cache.annotation.CacheExpiration
import org.goblinframework.core.util.DateUtils
import kotlin.math.max

/**
 * Calculate expiration seconds of specified [CacheExpiration] annotation.
 */
fun calculateExpirationInSeconds(annotation: CacheExpiration): Int {
  require(annotation.enable)
  return calculateExpirationInSeconds(annotation.policy, annotation.value)
}

/**
 * Calculate expiration seconds of specified [CacheExpiration.Policy] and value.
 * Return the specified value directly in case of policy is [CacheExpiration.Policy.FIXED]
 */
fun calculateExpirationInSeconds(policy: CacheExpiration.Policy, value: Int): Int {
  return when (policy) {
    CacheExpiration.Policy.FIXED -> max(0, value)
    CacheExpiration.Policy.TODAY -> DateUtils.getCurrentToDayEndSecond()
    CacheExpiration.Policy.THIS_WEEK -> DateUtils.getCurrentToWeekEndSecond()
    CacheExpiration.Policy.THIS_MONTH -> DateUtils.getCurrentToMonthEndSecond()
  }
}